ui/raidboss/data/07-dt/deepdungeon/pilgrims_traverse_the_final_verse.ts

```diff
@@ -24,25 +148,134 @@ const triggerSet: TriggerSet<Data> = {
     cn: '朝圣交错路 第99朝圣路/卓异的悲寂歼灭战',
   },
 
+  initData: () => ({
+    abyssalSides: false,
+  }),
+
   triggers: [
     // ---------------- Stone 99/The Final Verse Boss: Eminent Grief/Devoured Eater ----------------
+    {
+      id: 'PT 99 HP Difference Warning',
+      // 9F6 = Damage Up
+      // 105F = Rehabilitation
+      type: 'GainsEffect',
+      netRegex: { effectId: ['9F6', '105F'], capture: false },
+      suppressSeconds: 1,
+      alarmText: (_data, _matches, output) => output.text!(),
+      outputStrings: {
+        text: {
+          en: 'Check Boss HP Difference',
+        },
+      },
+    },
+    {
+      id: 'PT 99 Petrification/Hysteria',
+      // 01 = Petrification (failure from Light Vengeance expiring)
+      // 128 = Hysteria (failure from Dark Vengeance expiring)
+      type: 'GainsEffect',
+      netRegex: { effectId: ['01', '128'], capture: true },
+      infoText: (_data, matches, output) => {
+        const effect = matches.effect;
+        const target = matches.target;
+        return output.text!({ effect: effect, target: target });
+      },
+      outputStrings: {
+        text: {
+          en: '${effect} on ${target}',
+        },
+      },
+    },
+    {
+      id: 'PT 99 Light/Dark Vengeance Refresh Warning',
+      // 11CF = Dark Vengeance
+      // 11D0 = Light Vengeance
+      type: 'GainsEffect',
+      netRegex: { effectId: ['11CF', '11D0'], capture: true },
+      condition: Conditions.targetIsYou(),
+      preRun: (data, matches) => {
+        const timestamp = Date.parse(matches.timestamp);
+        const duration = parseFloat(matches.duration);
+        data.myVengeanceExpiration = timestamp + duration * 1000;
+      },
+      delaySeconds: (_data, matches) => parseFloat(matches.duration) - 10,
+      infoText: (data, matches, output) => {
+        const timestamp = Date.parse(matches.timestamp);
+        const duration = parseFloat(matches.duration);
+        const thisExpiration = timestamp + duration * 1000;
+        const myExpiration = data.myVengeanceExpiration;
+        if (myExpiration === undefined || myExpiration > thisExpiration)
+          return;
+        return output.text!();
```

Comment:
I don't know that we can do anything about this, but this doesn't work if the player gets a debuff prior to engaging because `data` gets re-initialized. I did specifically test gaining the effect in-combat and it does work there, but this is a limitation.

Reply:
I don't see a good way around this, so it's just going to have that limitation.